Sortable Widgets and Timelines

randompenguin1 opened the pull request:
The changes in this PR are related to the discussion in this thread: [https://forum.friendi.ca/display/373ebf56-7868-f87a-02b8-4aa456938304](https://forum.friendi.ca/display/373ebf56-7868-f87a-02b8-4aa456938304)

I'm leveraging the the jQuery Sortable plugin that is _already_ present in Friendica, though I have no idea if or where it is normally loaded or used at: _/view/asset/es-jquery-sortable/source/js/jquery-sortable-min.js_

This turns the list under **Settings > Additional Features > Network Widgets**  into a drag-n-drop sortable list to change the order in which they appear in the sidebar on the Network page.

The same method is also applied to the list under **Settings > Display > Timelines** which is changed from a table to two sortable lists, one for the order they appear in the "Channels" sidebar widget and another for the "Top Menu" on the Network page.

They can also be sorted via keyboard control by tabbing into the list and using the up/down arrows to move the item in the list. For screen readers there is also feedback about the action and position.

Drag-and-Drop does not work very well on touch devices, of course. When you try to drag on a phone or tablet it thinks you want to scroll. I included a compatibility script that at least makes it work on touch devices, though it's still a terrible experience and I wouldn't recommend sorting the lists that way. But that's a touchscreen limitation.

Here is a gif of a screen recording of it in action:

![Screen Recording 2025-10-28 at 9 29 43 PM](https://github.com/user-attachments/assets/e0f79db0-b18c-44b1-b547-0bd809f95e8e)

randompenguin1 committed:
Changes for sortable timelines

randompenguin1 committed:
Sortable timelines

Swaps table for sortable lists

randompenguin1 committed:
Changes for sortable widgets

randompenguin1 committed:
Scripts and styling for sortable lists

randompenguin1 committed:
Changes for sortable widgets

randompenguin1 committed:
fixed template array vars

accidentally added array items instead of editing values of existing ones.

annando commented on src/Module/Conversation/Network.php:
```diff
@@ -167,37 +224,25 @@ protected function content(array $request = []): string
+				"channels",
+				"searches",
+				"folders",
+				"nosharer",
```
"nosharer" should be invisible on some channels. How is this controlled now?

randompenguin1 replied:
It should still be controlled exactly as it was. I didn't change any of the code for displaying them, I just had to move it into a sub-function so I could call it in a different order if necessary.

annando commented on src/Module/Conversation/Network.php:
```diff
@@ -167,37 +224,25 @@ protected function content(array $request = []): string
+
+		if (empty($widgetorder)){
+			$widgetorder = [
+				"circles",
```
Is it needed to have the widgets specified as strings? Wouldn't (numerical) constants a better idea?

randompenguin1 replied:
Yes, they need to be specified as strings. And these are not the localized strings but based on the ID of the widget in the HTML code. And that array is just the fallback position if no custom order is set.

annando replied:
When they need to be strings it would be great if you could create constants for this anyway. That's much nicer to follow.

marcusxss replied:
@randompenguin1 I can push a commit that adds constants for these if you want?
Essentially adding:
```php
public const CIRCLES       = "circles";
public const GROUPS        = "forumlist_profile";
public const ARCHIVE       = "archive";
...
```
...at the top of the class and then refer to those in the switch/case and the widget order further down..

Or at least I think that's what's suggested?

marcusxss replied:
Looking closer at it, they all appear to be defined as constants in Content/Feature.php alredy, which are already being used by the conditionals - so perhaps we just refer to those?

marcusxss replied:
Could one omit the `widgets` method and rewrite the `foreach` in `content` that previously called it to something like this?:
```php
		foreach ($widgetorder as $widget) {
			if (Feature::isEnabled($this->session->getLocalUserId(), $widget)) {
				switch ($widget) {
					case Feature::CIRCLES:
						$this->page['aside'] .= Circle::sidebarWidget($module, $module . '/circle', 'standard', $this->circleId);
						break;
					case Feature::GROUPS:
						$this->page['aside'] .= GroupManager::widget($this->session->getLocalUserId());
						break;
					case Feature::ARCHIVE:
						$this->page['aside'] .= Widget::postedByYear($module . '/archive', $this->session->getLocalUserId(), false);
						break;
					case Feature::NETWORKS:
						$this->page['aside'] .= Widget::networks($module, $this->network);
						break;
					case Feature::ACCOUNTS:
						$this->page['aside'] .= Widget::accountTypes($module, $this->accountTypeString);
						break;
					case Feature::CHANNELS:
						$this->page['aside'] .= Widget::channels($module, $this->selectedTab, $this->session->getLocalUserId());
						break;
					case Feature::SEARCHES:
						$this->page['aside'] .= Widget\SavedSearches::getHTML($this->args->getQueryString());
						break;
					case Feature::FOLDERS:
						$this->page['aside'] .= Widget::fileAs('filed', '');
						break;
					case Feature::TRENDING_TAGS:
						$this->page['aside'] .= TrendingTags::getHTML($this->selectedTab);
						break;
					case Feature::NOSHARER:
						if (($this->channel->isTimeline($this->selectedTab) || $this->userDefinedChannel->isTimeline($this->selectedTab, $this->session->getLocalUserId())) &&
							!in_array($this->selectedTab, [Channel::FOLLOWERS, Channel::FORYOU, Channel::DISCOVER])) {
							$this->page['aside'] .= $this->getNoSharerWidget('network');
						}
						break;
				}
			}
		}
```

Edit: Updated to simplify it a bit further.

marcusxss replied:
I can update the PR with that version if wanted?

marcusxss replied:
I did that now.

annando commented:
Thanks for the PR! It really looks promising! But please have got a look at my comments and fix the code style errors.

randompenguin1 commented:
@annando  I have no idea what I'm supposed to fix for "code standards." I can't make sense of what it's trying to tell me to fix? I'm not really a programmer. All I can tell you is I tested my code thoroughly, it may not be pretty but it works.

annando commented:
@randompenguin1 no problem. Just execute `bin/dev/php-cs-fixer/vendor/bin/php-cs-fixer fix ...` from your command line with `...´ being replaced by the files that the tests are complaining about.

randompenguin1 commented:
Do what? How? I'm only using GitHub's web UI. I don't know anything about using the command line stuff or even if there's a way to do that from the web UI.

annando commented:
You don't have a local installation?

randompenguin1 commented:
@annando Nope. No local GitHub installation. I've only used the web UI and I don't know GitHub very well. My background is mostly as a graphic designer along with front-end web design/dev stuff. I've only picked up some PHP and jQuery over the years because of WordPress projects. So I'm a little out of my element here.

annando commented:
When you click on https://ci.friendi.ca/repos/2/pipeline/5276/16 and you scroll to the bottom, you see something like this:
<img width="834" height="375" alt="image" src="https://github.com/user-attachments/assets/3e5bfe99-8fbb-4a02-b9b0-24e1d41295c5" />
The lines with `-` are the existing ones, the `+` lines are the corrected ones. You can copy the whole log to some text editor and go through it line by line.

marcusxss commented:
This is really great! I specifically had a user who said changing the order of those was what he viewed as one of the main things that he thought should be changed to improve the user experience.

Since translations are touched `bin/run_xgettext.sh` also needs to be run. But I believe maintainers (like annando and nupplaphil) can also do that if they're given permission when the PR is created?

Also if new translations are added (as opposed to reusing existing ones) I don't know if it can make the RC, of if it should target `develop`?

Anyway I think this is not nearly as complex as learning PHP or JQuery so I think you can learn it but it will take some time. Maybe one good approach would be video tutorials. And fx. I will also try to help when needed.
I think the advantage is that if you're also working in core like this, you can make a lot more changes rather than having to hack around things or outright not being able to change them.

You can't have a local `github` installation so I think what is meant is whether you have a local friendica installation? Because if you do, inside that you have a bin directory where fx. `run_xgettext.sh` is, which is run when translations are updated.

You can have a local `git` installation, though, which can fx. interact with both github and gitlab to create branches here, which you can then make PRs (GitHub) or MRs (Gitlab) from. Essentially you can make all the changes locally in any text editor, and then use the git command line tool to make branches and commits and send them to github or gitlab.

But exactly how that's setup also depends on what OS you're using. I'm decently familiar with Windows and more so Linux, but macOS I'm not, so I'm not sure there. But macOS does have the benefit of being a lot more similar to Linux than Windows  under the hood, so that may make some things easier there. Fx. one may just be able to run run_xgettext.sh from macOS's default terminal as it should be able to run bash scripts natively. Whether Windows can do that I'm not quite sure, but certainly it's possible using something like Git Bash. Essentially if you install git for windows it may include a terminal and shell as well?

I've been doing some documentation cleanup recently (mostly a lot of manual copy-pasting from nupplaphils PR), and this makes me think about how maybe the documentation on how to setup a development environment could be better. My own dev setup is manually hacked together in many ways as I couldn't get the dev setups working out of the box (neither vagrant nor docker), so I probably use that guide myself to improve my dev setup, if someone more familiar with it updated it.

randompenguin1 committed:
code standards fixes

randompenguin1 commented:
@annando Okay, I went through an manually fixed the code standards fails, which were all just subjective spacing differences.

I have no clue what that **messages.po_check** is trying to tell me. I don't know what that even is.

randompenguin1 commented:
@mfxa Yes I have a local Friendica installation. Two in fact, one for stable and one for a recent build of the RC. They're running in XAMPP not Docker. Those are just so I can test Bookface changes before I make those edits in my GitLab repos. But this is all getting way beyond me and if it's too complicated to contribute to a project I probably just won't bother anymore. I don't enjoy coding.  I'd rather be drawing.

marcusxss commented:
@randompenguin1 
Alright! If it would help you, you can add me as a `Collaborator` to your fork of the friendica repo here on GitHub, then I could sometimes do things like run PHP CS Fixer (code standards), fix the messages error or rebases when needed?
It should be possible to add collaborators here:
https://github.com/randompenguin1/friendica/settings/access
Then I'll fix the messages issue?

Some explanation:

For the code standards, the idea is that if each user subjectively decides how to write things you end up with a code base with greatly varying coding practices. So it enforces a certain subjective standard that all changes must follow.
Fortunately with the command line tools, these "fixes" can be applied automatically. There are also checks that aren't related to style but logic-related, when things could fx. fail under certain circumstances. It's currently just done on the PHP files I believe, but similar things exist for e.g. CSS, HTML and JavaScript.

messages.po is a file that contains all the text on friendica, and what each of them should be translated into depending on a user's language setting. When changing or adding translations, the main main messages.po file, used as the basis for all translations, needs to be updated, as otherwise people are unable to make translations for the new/changed text, which is done on Transifex.
The main messages.po file is updated "automatically" when one runs the program, in the bin directory under the friendica directory, called `run_xgettext.sh`.

randompenguin1 commented:
@mfxa  Ok, I think I did the collaborator invite right. Thank you for the offer to help. I find GitHub very confusing and not very user friendly to casual coders who don't live and breathe this stuff.

marcusxss commented:
Glad to help! Yeah it takes time to learn! Hopefully we're fully on Forgejo (git.friendi.ca) soon, and hopefully it's better in this regard as well. I find it a bit hard to compare personally.

Updated "messages" so that check no longer fails.

randompenguin1 committed:
:last-of-type to :nth-last-of-type(2)

:last-of-type would work if list was ul > li but it is div > div and :last-of-type goes by element not class

randompenguin1 commented:
@mfxa - realized I need to fix one more thing in _/settings/head.tpl_ and it looks like it failed another check after I committed the change.

My fix was for the inline stylesheet to round the bottom corners of the lists. The way I'd originally coded it with pseudo-class `:last-of-type` would work if it was a `<ul>` with `<li>` elements inside, but the structure is actually `<div>` with more `<div>` elements and `:last-of-type` goes by element not class so I had to switch it to `:nth-last-of-type(2)` to target the actual last list element.

marcusxss commented:
Looking at it, it looks like the failing test is a false positive. Ie. it's the CI itself that failed, not your change. Confusingly it happens sometimes.

Edit: Most of the time when it's messages or code standards it's not a false positive, but some of the other tests do fail by themselves sometimes.

randompenguin1 committed:
added empty div fix

Makes sure last item in sortable widgets list has rounded bottom corners. Makes sure any floated elements above it get cleared.

marcusxss committed:
Update messages, code style

marcusxss commented:
Rebased and updated messages again after some merged PRs also made changes to messages.
Part of the CI is failing but it appears to be false positives.

I'm still unsure whether PRs that add or modify translations should target RC or develop currently, depending on whether the release is planned for just around the corner, so people don't have time to translate if new things are added, or if it's next month or something?

marcusxss committed:
Use constants from Feature.php

nupplaphil commented on src/Content/Widget.php:
```diff
@@ -608,6 +608,19 @@ public static function channels(string $base, string $channelname, int $uid): st
 			}
 		}
 
+		$widget_timelineorder = json_decode(DI::pConfig()->get($uid, 'system', 'widget_timeline_order'));
+		if (!empty($widget_timelineorder)) {
+			$tmp = [];
+			foreach($widget_timelineorder as $order) {
+				foreach($channels as $channel) {
+					if($channel['ref'] == $order) {
+						$tmp[] = $channel;
+					}
+				}
+			}
+			$channels = $tmp;
+		}
+
```
```suggestion
              $widget_timelineorder = DI::pConfig()->get($uid, 'system', 'widget_timeline_order');
              $widget_timelineorder = json_decode($widget_timelineorder, true);

              if (is_array($widget_timelineorder) && !empty($widget_timelineorder)) {
                    // Indizite channel per 'ref' 
                    $channelIndex = [];
                    foreach ($channels as $channel) {
                        if (!isset($channel['ref'])) {
                            continue;
                         }
                        $channelIndex[$channel['ref']] = $channel;
                    }

                    $orderedChannels = [];

                    foreach ($widget_timelineorder as $orderRef) {
                        if (isset($channelIndex[$orderRef])) {
                            $orderedChannels[] = $channelIndex[$orderRef];
                            unset($channelIndex[$orderRef]); 
                        }
                  }

                  // any leftovers
                  if (!empty($channelIndex)) {
                      $orderedChannels = array_merge($orderedChannels, array_values($channelIndex));
                  }

                  $channels = $orderedChannels;
              }
```

marcusxss replied:
Is this ready for just applying as is?

marcusxss replied:
@nupplaphil ? :-)

nupplaphil commented on src/Module/Conversation/Network.php:
```diff
@@ -148,6 +148,64 @@ public function __construct(
 		$this->userDefinedChannel = $userDefinedChannel;
 	}
 
+	protected function widgets(string $widget)
```
Can you add a short phpdoc plz about what the function is doing and the params

randompenguin1 replied:
Do you mean add a comment ahead of  it? Something like:

```
	/**
	 * Return a network widget for the currently logged-in user in their preferred order
	 *
	 * @param string $widget from $widgetorder array
	 * @return string
	 */
```

randompenguin1 replied:
Reading below it looks like @mfxa is reworking that function anyway. I'm kind of losing the plot on where this is at now.

marcusxss replied:
Yes, currently the method is gone again, so...?

nupplaphil commented on src/Module/Conversation/Network.php:
```diff
@@ -148,6 +148,64 @@ public function __construct(
 		$this->userDefinedChannel = $userDefinedChannel;
 	}
 
+	protected function widgets(string $widget)
```
I'd prefer

```suggestion
	protected function gerWidgetContent(string $widget): string
```
 And return the translated widget string and add it to the `aside` array-entry as it makes it better readable I guess

marcusxss replied:
Right now it's no longer a method call so this can't really be done anymore. Do we want the method back?

randompenguin1 committed:
Removed link loading global.css for each notification

global.css is _already_ loaded in the page head. Loading it over and over again with each notification item overrrides theme styling on the /notifications/system page, which messes up the styling of the Notifications drop-down under the bell icon too (but only on this page). Affects ALL themes because currently *none* of them include an override for this template.

randompenguin1 reviewed (commented):
I'm not sure what I'm supposed to be reviewing but GitHub is bugging me so I looked at the code and so long as it works I'm fine with it.

marcusxss commented:
Well, I resolved the conflict, updated messages again and added my changes to use the constants, which also removed the `widgets` method, moving the logic to `content`. I also collapsed some commits into each other. 

If things still work as they should, I think the remaining suggestion that's still applicable is this one: https://github.com/friendica/friendica/pull/15284#pullrequestreview-3458075693, unless we want the method back.

Inside settings it seems my widget/timeline order is not the one shown after saving? It resets back to the default for me, both before and after my changes. But the channel order on /network seems to be updated just fine. It could be a local, unrelated issue though.

Edit: Specifically I made [these changes](https://github.com/friendica/friendica/pull/15284/commits/df0089984f063ec9fb96f02dca00db823fe4ce51)

annando commented on src/Module/Settings/Display.php:
```diff
@@ -153,7 +157,16 @@ protected function post(array $request = [])
 		$this->pConfig->set($uid, 'system', 'embed_remote_media', $embed_remote_media);
 		$this->pConfig->set($uid, 'system', 'embed_media', $embed_media);
-
+		if ($widget_timeline_reset == 1) {
```
You can simply write `if ($widget_timeline_reset) {`

marcusxss replied:
Done.

annando commented on src/Module/Settings/Display.php:
```diff
@@ -153,7 +157,16 @@ protected function post(array $request = [])
+		} else {
+			$this->pConfig->set($uid, 'system', 'widget_timeline_order', $widget_timelineorder);
+		}
+		if ($menu_timeline_reset == 1) {
```
Please change this to `if ($menu_timeline_reset) {` as well.

marcusxss replied:
Done.

annando commented on src/Module/Settings/Features.php:
```diff
@@ -35,7 +35,13 @@ protected function post(array $request = [])
-				$this->pConfig->set($this->session->getLocalUserId(), 'feature', substr($k, 8), (bool)$v);
+				if (substr($k, 8) == 'widgetorder') { // not boolean, stringified array
+					$this->pConfig->set($this->session->getLocalUserId(), 'feature', 'widgetorder', $v);
+				} elseif (substr($k, 8) == 'resetorder' && (bool)$v == 1) {
```
You can reduce `(bool)$v == 1` to `$v`.

marcusxss replied:
Done - leaving the remaining for randompenguin for now.

randompenguin1 replied:
I changed the (bool)$v on line 46 to just $v as well. Not sure a boolean check was necessary there.

annando commented on view/templates/settings/display.tpl:
```diff
@@ -33,28 +33,30 @@
+		{{/foreach}}
+		<div class="settings-submit-wrapper">
+			<input type="hidden" name="widget_timeline_reset" value="0"/>
+			<input type="checkbox" id="widget_timeline_reset" name="widget_timeline_reset" value="1"/> <label for="widget_timeline_reset">{{$reset_label}}</label>
```
I would like to stick with our existing checkbox template instead of hardcoding the HTML elements here. Is there something that the template can't do?

randompenguin1 replied:
I think the reason I didn't use the checkbox template is it is intended for building entries in the form body not the panel footer, which breaks the layout, and requires defining more template variables this function doesn't need _just_ to use the checkbox template.

randompenguin1 replied:
Okay, I added template variables for the reset checkboxes so they now use the checkbox field template.

annando commented on view/theme/frio/templates/settings/features.tpl:
```diff
@@ -20,12 +20,21 @@
 				</div>
 				<div id="{{$g}}-settings-content" class="panel-collapse collapse" role="tabpanel" aria-labelledby="{{$g}}-settings-title">
-					<div class="panel-body">
+					<div class="panel-body {{if $g == "network"}}network sortable{{/if}}">
```
I would like to have such checks in the code instead of the template. This here can easily be missed when changing some stuff. It would be better to just inject some new boolean template variable instead.

randompenguin1 replied:
It's because the template loop is for ALL features not just these new sortable lists, so it checks to see if it's doing the "Network" feature and only adds those things if it is. It was my solution to not having to create two separate loops.  And I don't really see the difference between checking if the key is "network" vs checking if some boolean is true or false.

annando replied:
When we define a new boolean variable that is passed to the template, then we can use a constant in the code instead of a hardcoded string. And by using a constant, it is easier to find, when there maybe are changes in the structure in the future.

randompenguin1 replied:
I don't really see how a boolean constant would work here though. It's not a constant condition. It only becomes true on one iteration of the loop, when it is processing the "network" key of the array.

annando replied:
I have missed that. But then the string "network" should be passed as a constant to the template.

randompenguin1 replied:
I added a constant for the key name and a template variable to check against that constant rather than check against the string directly. That string is now defined in a constant at the top of _Features.php_

annando reviewed (changes requested):
Please have got a look at my comments. These are all code style related change requests. If you need help, there, please tell and I guess you will get support here.

marcusxss committed:
Adjustments after review

annando commented on src/Module/Settings/Features.php:
```diff
@@ -35,7 +35,13 @@ protected function post(array $request = [])
 		foreach ($request as $k => $v) {
 			if (strpos($k, 'feature_') === 0) {
-				$this->pConfig->set($this->session->getLocalUserId(), 'feature', substr($k, 8), (bool)$v);
+				if (substr($k, 8) == 'widgetorder') { // not boolean, stringified array
```
Can we use a constant instead? Otherwise it is really hard to find in the code where this value is set. And since now we use the variable `$k` a lot in that loop, it might be an idea, to have some `$key = substr($k, 8);` after line 37 and to use it then in the loop instead of constantly repeating `substr($k, 8);`.

randompenguin1 replied:
Again, I don't understand how you'd use a constant here. It's _inside_ the loop. $k keeps changing on every iteration of the loop. It's never constant.

randompenguin1 replied:
after line 37 it now has `$key = substr($k, 8);` and then references `$key` inside the if/else instead of searching the substring over and over.

annando commented:
Has someone tested it thoroughly on their machine already?

randompenguin1 commented:
> Has someone tested it thoroughly on their machine already?

I just downloaded all the changed files in their current state and tested it on my dev server. The order under "Additional Features > Network Widgets" works as intended.

The one under "Display > Timelines" works to change the order on the front-end, but on reload the Settings screen no longer shows them in your custom order so if you save settings on that page again it reverts them to the default order. Also ticking either of the "Reset" checkboxes will reset _both_ the Channels Widget and Top Menu. I'm not sure what changes broke that and I doubt I'll have time to work on it before the end of the year.

randompenguin1 committed:
Update Display.php

Added missing sortable list processing.

Switched template vars so resets now use checkbox template.

randompenguin1 committed:
changed reset checkboxes to use checkbox template

randompenguin1 committed:
reset checkboxes now use checkbox template

randompenguin1 committed:
Update Features.php

* switched key name substring check to $key variable
* added template variables for reset checkbox

randompenguin1 committed:
Update features.tpl

Changed so reset checkbox uses checkbox template.
$key is now checked against a constant instead of string

randompenguin1 committed:
added template variable for sort key check against constant string

randompenguin1 committed:
reset checkbox now uses checkbox template

randompenguin1 committed:
fixed css so reset checkbox field does not adopt sortable list styling

randompenguin1 committed:
(bool)$v changed to just $v

randompenguin1 commented:
Due to weather changing some of my holiday plans I did have time to look at this today and discovered the reason the _Display > Timelines_ were not working is that I had forgotten to add the actual processing code for the custom order of the sortable lists!  With that added to **Display.php** it now works as intended (that section of code - lines 297-363 - is heavily commented as to what each step is doing).

annando commented:
Thanks for your work. Can someone check it out on their system to test the functionality? There are still some things in the code that I would like to be changed, but this can be done afterwards as well. This PR already took a long time.

annando commented:
@randompenguin1 @mfxa please have got a look at the merge conflict.

marcusxss committed:
Messages, code style

marcusxss commented:
I haven't done much testing of it personally (yet), but I did resolve the conflicts and update messages.
Also rebased it on RC and undid the merge of RC into this branch.

sekretaerbaer commented:
I tested it once. The settings are applied and also displayed in the widget area. This applies to both the standard channels and the channels you have created yourself.
In the top menu area, the settings are also saved and applied as long as they concern the standard channels. However, this does not apply to channels you have created yourself; the settings are saved but not displayed.

<img width="862" height="383" alt="Bildschirmfoto 2025-12-25 um 05 51 30" src="https://github.com/user-attachments/assets/ad94f2eb-f3ae-4b03-a163-c53426e3eae8" />
<img width="862" height="111" alt="Bildschirmfoto 2025-12-25 um 05 52 03" src="https://github.com/user-attachments/assets/65c8e29a-7288-4b0a-b31f-c4a3216d3e62" />

annando commented:
@sekretaerbaer thanks for the testing!

sekretaerbaer commented:
I don't know if it has anything to do with the changes, but after I undid the pull (without resetting the order), I always ran into a timeout when logging in with the user, and the entire instance was no longer accessible.

randompenguin1 committed:
Fix for top menu tabs with custom channels

With custom channels the array key does not match the tab code string so now it checks for a match against either.

randompenguin1 commented:
@sekretaerbaer You were correct that it was not handling custom channels properly in the top menu tabs. I took a look at it and it's because the array key for a custom channel does not match the sub-key "code" value. So I added a check for a match against either and now, at least for me, it is showing the custom channels and in my chosen order in the Top Menu.

sekretaerbaer commented:
@randompenguin1 Everything is working now. Thank you for your work.

annando commented:
Since it seems that it is working now, I think that we can merge it now.

Tealk commented:
It works very well so far, thank you very much.
